warehouse/oidc/models/activestate.py

```diff
@@ -0,0 +1,204 @@
+
+    components = signed_claim.split(":")
+
+    if len(components) < 4:
+        raise InvalidPublisherError("Invalid 'subject' claim. Wrong format")
+
+    matches = (
+        f"{components[0]}:{components[1]}:{components[2]}:{components[3]}"
+        == ground_truth
+    )
+    if not matches:
+        raise InvalidPublisherError("Invalid 'subject' claim")
+
+    return True
+
+
+class ActiveStatePublisherMixin:
+    """
+    Common functionality for both pending and concrete ActiveState OIDC publishers.
+    """
+
+    organization = mapped_column(String, nullable=False)
+    activestate_project_name = mapped_column(String, nullable=False)
+    actor = mapped_column(String, nullable=False)
+    # 'actor' (The ActiveState platform username) is obstained from the user
+    # while configuring the publisher We'll make an api call to ActiveState to
+    # get the 'actor_id'
+    actor_id = mapped_column(String, nullable=False)
+    ingredient = mapped_column(String, nullable=True)
+
+    __required_verifiable_claims__: dict[str, CheckClaimCallable[Any]] = {
+        "sub": _check_sub,
+        "organization": oidccore.check_claim_binary(str.__eq__),
+        "project": oidccore.check_claim_binary(str.__eq__),
+        "actor_id": oidccore.check_claim_binary(str.__eq__),
+        # This is the name of the builder in the ActiveState Platform that
+        # publishes things to PyPI.
+        "builder": oidccore.check_claim_invariant("pypi-publisher"),
+    }
+
+    __optional_verifiable_claims__: dict[str, CheckClaimCallable[Any]] = {
+        # Should we check that the pypi projectname matches this ingredient
+        # name?
+        "ingredient": oidccore.check_claim_binary(str.__eq__),
+    }
+
+    __unchecked_claims__ = {
+        "actor",
+        "artifact_id",
+        "organization_id",
+        "project_id",
+        "project_path",
+        "project_visibility",
+    }
+
+    @staticmethod
+    def __lookup_all__(klass, signed_claims: SignedClaims):
+        return Query(klass).filter_by(
+            organization=signed_claims["organization"],
+            activestate_project_name=signed_claims["project"],
+            actor_id=signed_claims["actor_id"],
+        )
+
+    __lookup_strategies__ = [
+        __lookup_all__,
+    ]
+
+    @property
+    def sub(self) -> str:
+        return f"org:{self.organization}:project:{self.activestate_project_name}"
+
+    @property
+    def builder(self) -> str:
+        return "pypi-publisher"
+
+    @property
+    def publisher_name(self) -> str:
+        return "ActiveState"
+
+    @property
+    def project(self) -> str:
+        return self.activestate_project_name
+
+    @property
+    def activestate_project_path(self) -> str:
+        return f"{self.organization}/{self.activestate_project_name}"
+
+    def publisher_url(self, claims: SignedClaims | None = None) -> str:
+        return urllib.parse.urljoin(ACTIVESTATE_URL, self.activestate_project_path)
```

Comment:
```suggestion
    def publisher_url(self, claims: SignedClaims | None = None) -> str:
        return urllib.parse.urljoin(
            ACTIVESTATE_URL,
            f"{self.organization}/{self.activestate_project_name}",
        )
```
The `activestate_project_path` property doesn't seem to be used anywhere else.

Reply:
Agreed.  Done locally.